R/coherence.R

```diff
@@ -189,96 +195,157 @@ coherence =  function( top_term_matrix
-    #                   } else {
-    #                     coh = as.data.table(term_index_combinations(idxs, comb_type = "one_idx-preceeding_idxs-topic_order", topic_order = restore_topic_order))
-    #                     coh[, score:= mapply(function(x,y) {log(.01 + twcm[x,y]) - log(.01 +twcm[y,y])}
-    #                                          ,V1, V2)]
-    #                     return(coh[, round(sum(score, na.rm = T), d = 4)])
-    #                   }
-    # })]
-#}
-
-#if ("UMass_stm_like_uppertri" %in% measure | measure == "all") {
-    # coherence[,sum_logratio_stmlike_uppertri:= sapply(twcm_term_idxs, function(idxs) {
-    #   if (length(idxs) < 2) {
-    #     return(NA)
-    #   } else {
-    #     reorder = order(match(idxs, restore_topic_order), decreasing = TRUE)
-    #     idxs = idxs[reorder]
-    #     res = twcm[idxs, idxs] + .01
-    #     res = apply(res, 2, function(x) x/(diag(res)))
-    #     res = log(res[upper.tri(res)])
-    #     return(round(sum(res, na.rm = T), d = 4))
-    #   }
-    # })]
-#}
-
-
-#if ("UMass_stm_like_lowertri" %in% measure | measure == "all") {
-    # coherence[,sum_logratio_stmlike_lowertri:= sapply(twcm_term_idxs, function(idxs) {
-    #   if (length(idxs) < 2) {
-    #     return(NA)
-    #   } else {
-    #     reorder = order(match(idxs, restore_topic_order), decreasing = TRUE)
-    #     idxs = idxs[reorder]
-    #     res = twcm[idxs, idxs]
-    #     res = apply(res, 2, function(x) x/(diag(res)+.01))
-    #     res = log(.01 + res[lower.tri(res)])
-    #     return(round(sum(res, na.rm = T), d = 4))
-    #   }
-    # })]
-#}
-
-
-    #Type 2
-    #mean(..., na.rm = T) for aggregation
-    #one_pre_topic_order as term indices
-#if ( "UMass_Torder" %in% measure | measure == "all") {
-    # coherence[,mean_prob_logratio_Torder:= sapply(twcm_term_idxs, function(idxs) {
-    #                 if (length(idxs) < 2) {
-    #                   return(NA)
-    #                 } else {
-    #                   coh = as.data.table(term_index_combinations(idxs, comb_type = "one_idx-preceeding_idxs-topic_order", topic_order = restore_topic_order))
-    #                   coh[, score:= mapply(function(x,y) {log(log_smooth_constant + twcm[x,y]) - log(twcm[y,y])}
-    #                                        ,V1, V2)]
-    #                   return(coh[, round(mean(score, na.rm = T), d = 4)])
-    #                 }
-    #   })]
-#}
-
-# #if ( "pmi" %in% measure | measure == "all") {
-    coherence[,mean_pmi:= sapply(twcm_term_idxs, function(idxs) {
+
+  if ("mean_logratio" %in% measure | "all" %in% measure) {
+    coherence[,mean_logratio:= sapply(twcm_term_idxs, function(idxs) {
+      if (length(idxs) < 2) {
+        return(NA)
+      } else {
+        reorder = order(match(idxs, restore_topic_order), decreasing = TRUE)
+        idxs = idxs[reorder]
+        res = twcm[idxs, idxs]
+        res[upper.tri(res)] = res[upper.tri(res)] + log_smooth_constant
+        res = res/rev(diag(res))
```

Comment:
@manuelbickel  I've see `res/rev(diag(res))` in several places, but I'm not sure I understand why is it like this. Could you please clarify a bit?

Reply:
When I tried to vectorize the `mapply` loops, e.g., for pmi `coh[, score:= mapply(function(x,y)  {log2((twcm[x,y]/n_twcm_windows) + log_smooth_constant) - log2(twcm[x,x]/n_twcm_windows) - log2(twcm[y,y]/n_twcm_windows)} ,V1, V2)]` where `x` and `y` represent an index pair of the `tcm` I simply played around with different options to reproduce the result in a vectorized version. This division represents the part `/twcm[y,y]`. I have to admit it does seem strange that this works, but with this approach I could reproduce all numbers. I also tried ´apply(res, 1, function(x) res/diag(x))´  which did reproduce the numbers. Maybe there is a more intuitive way, but I could not wrap my head around the logic of index ordering and subsetting in a better way. Does that help?

Reply:
there was an error in above message, `apply` did not reproduce numbers...

Reply:
I think I made a mistake, thanks for highlighting this. [This approach](https://gist.github.com/manuelbickel/7ab6b96845e99e6ea471caf06af530dd) should be closer to what is needed.

Reply:
Think I found the error why my NMPI vectorization did not work. The lines you highlighted need to be replaced by (I forgot to use `t()` on the result of `apply` to restore the original matrix structure):
`#res = res/rev(diag(res)) should read something like` 
`d = diag(res)`
`res = t(apply(res, 1, function(x) res/d))`